### PR TITLE
Speedup mdtest parser

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

A few small changes to speed up the mdtest parser:

* Use `memchr` to find possible header and codeblock positions over matching both regexes (fewer LazyLock lookups, skip over more text than line by line)
* Avoid use of `captures` for the header (because captures are "slow" and header parsing is easy)
* Avoid using unicode words for the markdown language

## Test Plan

I "benchmarked" the parser by commenting out the `run_test` invocation. 

Before: ±230ms
Now: ±170ms

### CODE REVIEW

**Review by AlexWaygood — APPROVED** [review]
This all looks fine to me but @carljm is obviously the expert on this code. I haven't benchmarked myself but I trust you when you say it's faster 😄

File: crates/red_knot_test/src/parser.rs [hunk 1]
```diff
@@ -136,12 +137,13 @@ pub(crate) struct EmbeddedFile<'s> {
 /// Matches an arbitrary amount of whitespace (including newlines), followed by a sequence of `#`
 /// characters, followed by a title heading, followed by a newline.
 static HEADER_RE: LazyLock<Regex> =
-    LazyLock::new(|| Regex::new(r"^(\s*\n)*(?<level>#+)\s+(?<title>.+)\s*\n").unwrap());
+    LazyLock::new(|| Regex::new(r"^(?<level>#+)\s+(?<title>.+)\s*\n").unwrap());
```
Comment (sharkdp) [inline crates/red_knot_test/src/parser.rs:140]: Minor: the doc-comment should be updated ("Matches an arbitrary amount of whitespace"…)

File: crates/red_knot_test/src/parser.rs [hunk 2]
```diff
@@ -136,12 +137,13 @@ pub(crate) struct EmbeddedFile<'s> {
...
 
 /// Matches a code block fenced by triple backticks, possibly with language and `key=val`
 /// configuration items following the opening backticks (in the "tag string" of the code block).
 static CODE_RE: LazyLock<Regex> = LazyLock::new(|| {
-    Regex::new(r"^```(?<lang>\w+)(?<config>( +\S+)*)\s*\n(?<code>(.|\n)*?)\n?```\s*\n").unwrap()
+    Regex::new(r"^```(?<lang>(?-u:\w)+)(?<config>( +\S+)*)\s*\n(?<code>(.|\n)*?)\n```\s*\n")
```
Comment (sharkdp) [inline crates/red_knot_test/src/parser.rs:145]: The new regex does not match empty code blocks such as
````markdown
```py path=a.py
```
````
in `mdtest/import/errors.md`, which is also why some tests are failing. The final newline was optional before: `\n?`.

Comment (AlexWaygood) [reply crates/red_knot_test/src/parser.rs]: It might be worth adding a specific unit test for this edge case in the `red_knot_test` crate, so that we still have tests that will fail even if the mdtest in `mdtest/import/errors.md` changes so it no longer has an empty code block

Comment (MichaReiser) [reply crates/red_knot_test/src/parser.rs]: Agree, I'll follow up, once I have @carljm approval

File: crates/red_knot_test/src/parser.rs [hunk 3]
```diff
@@ -226,33 +227,60 @@ impl<'s> Parser<'s> {
...
+                match c {
+                    '#' => {
+                        if let Some(find) = HEADER_RE.find(self.unparsed) {
+                            self.parse_header(find.as_str())?;
+                            self.unparsed = &self.unparsed[find.end()..];
+                            continue;
+                        }
+                    }
+                    '`' => {
+                        if let Some(captures) = self.scan(&CODE_RE) {
```
Comment (carljm) [inline crates/red_knot_test/src/parser.rs:247]: I don't think it's worth having a separate `scan` method with only one call site; it's quite short, let's just inline it here. That makes the behavior more parallel to the `HEADER_RE` case (the updating of `self.unparsed` is visible in both branches), which I think makes it easier to understand the logic.

File: crates/red_knot_test/src/parser.rs [hunk 4]
```diff
@@ -133,15 +134,15 @@ pub(crate) struct EmbeddedFile<'s> {
     pub(crate) code: &'s str,
 }
 
-/// Matches an arbitrary amount of whitespace (including newlines), followed by a sequence of `#`
-/// characters, followed by a title heading, followed by a newline.
+/// Matches a sequence of `#` characters, followed by a title heading, followed by a newline.
 static HEADER_RE: LazyLock<Regex> =
-    LazyLock::new(|| Regex::new(r"^(\s*\n)*(?<level>#+)\s+(?<title>.+)\s*\n").unwrap());
+    LazyLock::new(|| Regex::new(r"^(?<level>#+)\s+(?<title>.+)\s*\n").unwrap());
```
Comment (carljm) [inline crates/red_knot_test/src/parser.rs:139]: Do we need the capture groups in this regex if they are not used?

File: crates/red_knot_test/src/parser.rs [hunk 5]
```diff
@@ -428,6 +455,57 @@ mod tests {
...
+
+        let [test1, test2] = &mf.tests().collect::<Vec<_>>()[..] else {
+            panic!("expected two tests");
+        };
+
+        assert_eq!(test1.name(), "file.md - One");
+        assert_eq!(test2.name(), "file.md - Two");
+
+        let [main, foo] = test1.files().collect::<Vec<_>>()[..] else {
+            panic!("expected two file");
```
Comment (carljm) [inline crates/red_knot_test/src/parser.rs:489]:
```suggestion
            panic!("expected two files");
```

**Review by carljm — APPROVED** [review]
Looks great, thank you!!

File: crates/red_knot_test/src/parser.rs (hunk 4 above)
Comment (MichaReiser) [reply crates/red_knot_test/src/parser.rs:139]: Technically not but there should be no cost because we use `matches`. I do like them for documentation purposes.
